### PR TITLE
Introduce SymbolicValue in SE

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
`Squash ExplodedNode` commit doesn't need a review

### CODE REVIEW

**Comment by pavel-mikula-sonarsource** [comment]
cc @costin-zaharia-sonarsource

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs [hunk 1]
```diff
@@ -18,11 +18,25 @@
...
 {
     public sealed class ProgramState
     {
-        public static readonly ProgramState Empty = new ProgramState();
-        // ToDo: Some stuff is missing here
+        public static readonly ProgramState Empty = new ProgramState(ImmutableDictionary<IOperationWrapperSonar, SymbolicValue>.Empty);
+
+        private readonly ImmutableDictionary<IOperationWrapperSonar, SymbolicValue> operationValue;
+
+        public SymbolicValue this[IOperationWrapperSonar operation] => operationValue.TryGetValue(operation, out var value) ? value : null;
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs:32]: nitpick: public properties/accessor before private fields?

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs [hunk 2]
```diff
@@ -18,11 +18,25 @@
...
+
 namespace SonarAnalyzer.SymbolicExecution.Roslyn
 {
     public sealed class ProgramState
     {
-        public static readonly ProgramState Empty = new ProgramState();
-        // ToDo: Some stuff is missing here
+        public static readonly ProgramState Empty = new ProgramState(ImmutableDictionary<IOperationWrapperSonar, SymbolicValue>.Empty);
+
+        private readonly ImmutableDictionary<IOperationWrapperSonar, SymbolicValue> operationValue;
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs:30]: I think a comment explaining what this is will help - is the dictionary value the result of the operation?

or maybe rename to `operationResult` (and from the signature we can see the result is a SymbolicValue)?  I'd find this more intuitive

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs [hunk 3]
```diff
@@ -18,10 +18,25 @@
...
+
+        public SymbolicValue this[IOperationWrapperSonar operation] => operationValue.TryGetValue(operation, out var value) ? value : null;
+
+        private ProgramState(ImmutableDictionary<IOperationWrapperSonar, SymbolicValue> operationValue)
+        {
+            this.operationValue = operationValue;
+        }
+
+        public ProgramState AddOperationValue(IOperationWrapperSonar operation, SymbolicValue value) =>
+            new ProgramState(operationValue.Add(operation, value));
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs:40]: double-check : if the `key` already exists, it is an illegal operation and it's fine to let the dictionary throw?

please add a test for this scenario?

**Review by andrei-epure-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.SymbolicExecution.Roslyn
+{
+    public class SymbolicValue
+    {
+        private readonly int identifier;
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs:25]: This used to be `static int` int the old SE. Shared for the lifetime of the Roslyn instance. So we should be fine now 😆

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs:32]: This is correct. We should have all fields (public, protected, private) followed by all properties (public, protected, private).

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs:40]: Good catch

**Review by andrei-epure-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**